File: client/client/model/ExerciseSession.swift
import Foundation

enum ExerciseSessionOperationType: String, CaseIterable, Codable {
    case sessionJoin = "session_join"
    case sessionLeave = "session_leave"
    case sessionUpdate = "session_update"
    case sessionSync = "session_sync"
    
    case exerciseAdd = "exercise_add"
    case exerciseUpdate = "exercise_update"
    case exerciseDelete = "exercise_delete"
    
    case setAdd = "set_add"
    case setUpdate = "set_update"
    case setDelete = "set_delete"
    case setComplete = "set_complete"
    
    case cursorMove = "cursor_move"
    case syncRequest = "sync_request"
    case syncResponse = "sync_response"
}

enum ExerciseType: String, Codable {
    case weightReps = "weight_reps"
    case weightTime = "weight_time"
    case distanceTime = "distance_time"
    case reps = "reps"
    case time = "time"
    case distance = "distance"
}

enum ExerciseSessionStatus: String, Codable {
    case draft = "draft"
    case active = "active"
}

enum ExerciseSessionStateItemType: String, Codable {
    case single = "single"
    case compound = "compound"
}

enum ExerciseSetType: String, Codable {
    case warmupSet = "warmup"
    case workingSet = "working"
    case dropSet = "drop"
    case superSet = "super"
    case failureSet = "failure"
}

enum WeightUnit: String, Codable {
    case kilogram = "kg"
    case pound = "lb"
}

enum DistanceUnit: String, Codable {
    case meter = "m"
    case kilometer = "km"
    case mile = "mi"
    case yard = "yd"
}

struct Weight: Codable {
    var value: Double
    var unit: WeightUnit
    
    var toKg: Double {
        unit == .pound ? value * 0.453592 : value
    }
    
    var toPound: Double {
        unit == .kilogram ? value / 0.453592 : value
    }
}

struct Distance: Codable {
    var value: Double
    var unit: DistanceUnit = .meter
    
    var toMeters: Double {
        switch unit {
        case .meter: return value
        case .kilometer: return value * 1_000
        case .mile: return value * 1_609.34
        case .yard: return value * 0.9144
        }
    }
}

struct Duration: Codable {
    var value: Int
}

struct ExerciseSessionParticipantCursor: Codable {
    var exerciseId: String
    var exerciseSetId: String
    
    enum CodingKeys: String, CodingKey {
        case exerciseId = "exercise_id"
        case exerciseSetId = "exercise_set_id"
    }
}

struct ExerciseSessionParticipant: Codable {
    var id: String
    var color: String
    var cursor: ExerciseSessionParticipantCursor?
}

struct ExerciseSessionInvitation: Codable {
    var invitedBy: String
    var invited: String
    var expires: Date?
    
    enum CodingKeys: String, CodingKey {
        case invitedBy = "invited_by"
        case invited
        case expires
    }
}

struct ExerciseSession: Codable {
    var id: String
    var name: String?
    var status: ExerciseSessionStatus
    var ownerId: String
    var createdAt: Date
    var updatedAt: Date
    var participants: [ExerciseSessionParticipant]
    var invitations: [ExerciseSessionInvitation]
    
    enum CodingKeys: String, CodingKey {
        case id = "id"
        case name
        case status
        case ownerId = "owner_id"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case participants
        case invitations
    }
}

struct ExerciseSessionItemMeta: Codable {
    var internalId: String
    var name: String
    var type: ExerciseType
    
    enum CodingKeys: String, CodingKey {
        case internalId = "internal_id"
        case name
        case type
    }
}

struct ExerciseSessionStateItemMetric: Codable {
    var reps: Int?
    var weight: Weight?
    var duration: Duration?
    var distance: Distance?
}

struct ExerciseSessionStateItemSet: Codable, Identifiable {
    var id: String
    var order: Int
    var metrics: ExerciseSessionStateItemMetric
    var type: ExerciseSetType
    var complete: Bool
}

struct ExerciseSessionStateItem: Codable, Identifiable {
    var id: String
    var order: Int
    var participants: [String]
    var type: ExerciseSessionStateItemType
    var rest: Int?
    var meta: [ExerciseSessionItemMeta]
    var sets: [ExerciseSessionStateItemSet]
}

struct ExerciseSessionState: Codable {
    var sessionId: String
    var accountId: String
    var version: Int
    var items: [ExerciseSessionStateItem]
    
    enum CodingKeys: String, CodingKey {
        case sessionId = "session_id"
        case accountId = "account_id"
        case version
        case items
    }
}

struct ExerciseSessionMessage: WebSocketMessage, Codable {
    let id: String
    let type: ExerciseSessionOperationType
    let sessionId: String?
    let payload: [String: AnyCodable]
    let timestamp: Date
    let version: Int
    let correlationId: String?
    
    var action: String {
        return type.rawValue
    }
    
    enum CodingKeys: String, CodingKey {
        case id, type, payload, timestamp, version
        case sessionId = "session_id"
        case correlationId = "correlation_id"
    }
    
    init(
        id: String = UUID().uuidString,
        type: ExerciseSessionOperationType,
        sessionId: String? = nil,
        payload: [String: Any] = [:],
        version: Int = 0,
        correlationId: String? = nil
    ) {
        self.id = id
        self.type = type
        self.sessionId = sessionId
        self.payload = payload.mapValues { AnyCodable($0) }
        self.timestamp = Date()
        self.version = version
        self.correlationId = correlationId
    }
}

struct ExerciseSessionJoinMessage {
    static func create(sessionId: String) -> ExerciseSessionMessage {
        return ExerciseSessionMessage(
            type: .sessionJoin,
            sessionId: sessionId,
            payload: ["session_id": sessionId]
        )
    }
}

struct ExerciseSessionLeaveMessage {
    static func create(sessionId: String) -> ExerciseSessionMessage {
        return ExerciseSessionMessage(
            type: .sessionLeave,
            sessionId: sessionId,
            payload: ["session_id": sessionId]
        )
    }
}

File: server/app/schema/exercise_session.py
from typing import Optional, List
from enum import Enum
from bson import ObjectId
from datetime import datetime
from pydantic import BaseModel, Field, ConfigDict

# For more information about this schema, see:
# /server/schemas/exercise.json

class ExerciseSessionStatus(str, Enum):
    DRAFT = "draft"
    ACTIVE = "active"

class ExerciseSessionStateItemType(str, Enum):
    SINGLE = "single"
    COMPOUND = "compound"

class ExerciseType(str, Enum):
    WEIGHT_REPS = "weight_reps"
    WEIGHT_TIME = "weight_time"
    DISTANCE_TIME = "distance_time"
    REPS = "reps"
    TIME = "time"
    DISTANCE = "distance"

class ExerciseSetType(str, Enum):
    WARMUP = "warmup"
    WORKING = "working"
    DROP = "drop"
    SUPER = "super"
    FAILURE = "failure"

class WeightUnit(str, Enum):
    KILOGRAM = "kg"
    POUND = "lb"

class DistanceUnit(str, Enum):
    METER = "m"
    KILOMETER = "km"
    MILE = "mi"
    YARD = "yd"

class Weight(BaseModel):
    value: float
    unit: WeightUnit = WeightUnit.POUND

    def to_kg(self) -> float:
        return self.value * 0.453592 if self.unit == WeightUnit.POUND else self.value

    def to_lb(self) -> float:
        return self.value * 2.20462 if self.unit == WeightUnit.KILOGRAM else self.value

class Duration(BaseModel):
    value: int  # seconds

class Distance(BaseModel):
    value: float
    unit: DistanceUnit = DistanceUnit.METER

    def to_meters(self) -> float:
        conv = {
            DistanceUnit.METER: 1,
            DistanceUnit.KILOMETER: 1000,
            DistanceUnit.MILE: 1609.34,
            DistanceUnit.YARD: 0.9144,
        }
        return self.value * conv[self.unit]

class ExerciseSessionParticipantCursor(BaseModel):
    exercise_id: str
    exercise_set_id: str

class ExerciseSessionParticipant(BaseModel):
    id: str
    color: str
    cursor: Optional[ExerciseSessionParticipantCursor] = None

class ExerciseSessionInvitation(BaseModel):
    invited_by: str
    invited: str
    expires: Optional[datetime] = None

class ExerciseSession(BaseModel):
    name: Optional[str] = None
    status: ExerciseSessionStatus
    owner_id: str
    created_at: datetime
    updated_at: datetime
    participants: List[ExerciseSessionParticipant] = Field(default_factory=list)
    invitations: List[ExerciseSessionInvitation] = Field(default_factory=list)

class ExerciseSessionItemMeta(BaseModel):
    internal_id: str
    name: str
    type: ExerciseType

class ExerciseSessionStateItemMetric(BaseModel):
    reps: Optional[int] = None
    weight: Optional[Weight] = None
    duration: Optional[Duration] = None
    distance: Optional[Distance] = None

class ExerciseSessionStateItemSet(BaseModel):
    id: str
    order: int = 1
    metrics: ExerciseSessionStateItemMetric
    type: ExerciseSetType
    complete: bool = False

class ExerciseSessionStateItem(BaseModel):
    id: str
    order: int = 1
    participants: List[str] = Field(default_factory=list)
    type: ExerciseSessionStateItemType
    rest: Optional[int] = None
    meta: List[ExerciseSessionItemMeta] = Field(default_factory=list)
    sets: List[ExerciseSessionStateItemSet] = Field(default_factory=list)

    model_config = ConfigDict(populate_by_name=True)

class ExerciseSessionState(BaseModel):
    session_id: str
    account_id: str
    version: int = 0
    items: List[ExerciseSessionStateItem] = Field(default_factory=list)

class ExerciseSessionInDB(ExerciseSession):
    id: Optional[str] = Field(default=None, alias="_id")

    class Config:
        allow_population_by_field_name = True
        arbitrary_types_allowed = True
        json_encoders = {
            ObjectId: str,
        }